### PR TITLE
Coverage support with Cadence Tools

### PR DESCRIPTION
**Pull request by ctopal**
Enables coverage collection while running the ibex-dv with xcelium.

Signed-off-by: Canberk Topal <ctopal@lowrisc.org>

### CODE REVIEW

**Comment by ctopal** [comment]
This enables a coverage report to be generated from the runs but as far as I can tell from the reports, we are missing the functional coverage numbers. However, it at least allows us to improve the compatibility of Ibex DV with Cadence tools. 

@rswarbrick : What do you think?

**Comment by ctopal** [comment]
Last changes actually enables fcov collection, oops! Should be fine overall now (except for my not so pretty code)

File: dv/uvm/core_ibex/Makefile [hunk 1]
```diff
@@ -16,6 +16,10 @@ export LD_LIBRARY_PATH := $(IBEX_COSIM_ISS_ROOT)/lib/:${LD_LIBRARY_PATH}
 endif
 endif
 
+# Needed for tcl files that are used with Cadence tools.
+export dv_root	    := $(realpath ../../../vendor/lowrisc_ip/dv)
```
Comment (rswarbrick) [inline dv/uvm/core_ibex/Makefile:20]: Nit: Indentation (extra space before the colon?)

File: dv/uvm/core_ibex/sim.py [hunk 2]
```diff
@@ -101,30 +102,68 @@ def gen_cov(base_dir, simulator, lsf_cmd):
...
+        os.makedirs(report_dir, exist_ok=True)
+        # First do the merge
+        cov_cmd = "imc -64bit -licqueue -exec "
+        cov_cmd += _XCELIUM_SCRIPTS + "/cov_merge.tcl"
+        subprocess.run(cov_cmd, shell=True, env= my_env)
+        # Then do the reporting
+        cov_cmd = "imc -64bit -licqueue -load " + merge_db_dir + " -exec "
+        cov_cmd += _XCELIUM_SCRIPTS + "/cov_report.tcl"
+        subprocess.run(cov_cmd, shell=True, env= my_env)
+        run_cmd(cov_cmd)
```
Comment (rswarbrick) [inline dv/uvm/core_ibex/sim.py:166]: I think you're running the report command twice?

File: dv/uvm/core_ibex/sim.py [hunk 3]
```diff
@@ -101,30 +102,68 @@ def gen_cov(base_dir, simulator, lsf_cmd):
...
+            sys.exit(RET_SUCCESS)
+        # Set a common merge directory to set as a temporary environment variable
+        merge_dir = "out/cov_merge"
+        merge_db_dir = "out/cov_merge/merged"
+        report_dir = "out/cov_report"
+        # Copy our existing environment variables.
+        my_env = os.environ.copy()
+        # Variables below are needed for our tcl files. Temporarily add them into
+        # environment variables list.
+        my_env["cov_merge_dir"] = merge_dir
```
[the remark this answers is not in the document]
Comment (ctopal) [reply dv/uvm/core_ibex/sim.py:146]: Oh this looks so much better! Thanks for suggesting this.

**Comment by ctopal** [comment]
Thanks @rswarbrick for the suggestions and review! With the latest push, we now collect run specific coverage report (under `out/seed-XXXX/cov_report`. So it should be fine all around now.

File: dv/uvm/core_ibex/sim.py [hunk 4]
```diff
@@ -125,6 +132,35 @@ def gen_cov(base_dir, simulator, lsf_cmd):
...
+
+        # Create the mentioned directories.
+        for d in cov_dirs.values():
+            os.makedirs(d, exist_ok=True)
+
+        # The merge TCL code uses a glob to find all available scopes and
+        # previous runs. In order to actually get the databases we need to
+        # go up once so that the "*" finds the directory we've seen.
+        cov_dirs['cov_db_dirs'] = ' '.join([os.path.join(d, '..')
+                                 for d in dir_list])
```
Comment (rswarbrick) [inline dv/uvm/core_ibex/sim.py:151]: The indentation is a bit off here (`for` should line up with `os`)

**Review by rswarbrick — APPROVED** [review]
A couple of nits, but this looks great. Thanks!

**Pull request merged**